File: app/src/main/java/com/thssh/example/LoadMoreRecyclerView.java
package com.thssh.example;

import android.content.Context;
import android.util.AttributeSet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class LoadMoreRecyclerView extends RecyclerView {

    public interface OnRefreshListener {
        void onRefresh();
    }

    public interface OnLoadMoreListener {
        void onLoadMore();
    }

    public interface OnRequestListener extends OnRefreshListener, OnLoadMoreListener{

    }

    private OnLoadMoreListener loadMoreListener;
    private OnRefreshListener refreshListener;

    public void setOnLoadMoreListener(OnLoadMoreListener listener) {
        this.loadMoreListener = listener;
    }

    public void setOnRefreshListener(OnRefreshListener listener) {
        this.refreshListener = listener;
    }

    public void setOnRequestListener(OnRequestListener listener) {
        setOnLoadMoreListener(listener);
        setOnRefreshListener(listener);
    }

    public LoadMoreRecyclerView(@NonNull Context context) {
        super(context);
        initView(context, null, 0);
    }

    public LoadMoreRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initView(context, attrs, 0);
    }

    public LoadMoreRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initView(context, attrs, defStyle);
    }

    final class InnerScrollListener extends RecyclerView.OnScrollListener {

        private boolean isLoading;
        private int firstVisibleItemPosition, lastVisibleItemPosition;

        private RecyclerView.LayoutManager layoutManager;

        public RecyclerView.LayoutManager getLinearLayoutManager() {
            if (layoutManager == null) {
                layoutManager = getLayoutManager();
            }
            return layoutManager;
        }

        @Override
        public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
            RecyclerView.LayoutManager layoutManager = getLinearLayoutManager();
            if (newState == RecyclerView.SCROLL_STATE_IDLE && layoutManager != null) {
                int visibleItemCount = layoutManager.getChildCount();
                int totalItemCount = layoutManager.getItemCount();
                if (visibleItemCount < totalItemCount && !isLoading) {
                    if (firstVisibleItemPosition == 0) {
                        isLoading = true;
                        if (refreshListener != null) refreshListener.onRefresh();
                    }else if (lastVisibleItemPosition == totalItemCount - 1) {
                        isLoading = true;
                        if (loadMoreListener != null) loadMoreListener.onLoadMore();
                    }
                }
            }
        }

        @Override
        public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
            if (getLayoutManager() instanceof LinearLayoutManager) {
                lastVisibleItemPosition = ((LinearLayoutManager) getLayoutManager()).findLastVisibleItemPosition();
                firstVisibleItemPosition = ((LinearLayoutManager) getLayoutManager()).findFirstVisibleItemPosition();
            } else {
                firstVisibleItemPosition = -1;
                lastVisibleItemPosition = -1;
            }
        }
    }

    public void loadingCompleted() {
        getScrollListener().isLoading = false;
    }

    private InnerScrollListener mScrollListener;

    InnerScrollListener getScrollListener() {
        if (mScrollListener == null) {
            mScrollListener = new InnerScrollListener();
        }
        return mScrollListener;
    }


    private void initView(Context context, AttributeSet attrs, int defStyle) {
        addOnScrollListener(getScrollListener());
    }


}

File: recyclerview/src/main/java/com/thssh/recyclerview/LoadMoreRecyclerView.kt
package com.thssh.recyclerview

import android.content.Context
import android.util.AttributeSet
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class InnerScrollListener: RecyclerView.OnScrollListener() {

    var isLoading = false
        set(value) {
            if (!value) {
                viewStateListener?.invoke(false, false)
            }
            field = value
        }

    var firstVisiblePosition = -1
    var lastVisiblePosition = -1

    var onRefreshListener: (() -> Unit)? = null
    var onLoadMoreListener: (() -> Unit)? = null
    var viewStateListener: ((isRefresh: Boolean, isLoadMore: Boolean) -> Unit)? = null

    override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
        if (newState == RecyclerView.SCROLL_STATE_IDLE) {
            recyclerView.layoutManager.let {
                if (!isLoading && it != null && it.childCount < it.itemCount) {
                    if (firstVisiblePosition == 0) {
                        isLoading = true
                        viewStateListener?.invoke(true, false)
                        // refresh
                        onRefreshListener?.invoke()
                    } else if (lastVisiblePosition == it.itemCount - 1) {
                        isLoading = true
                        viewStateListener?.invoke(false, true)
                        // loadMore
                        onLoadMoreListener?.invoke()
                    }
                }
            }
        }
    }

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        if (recyclerView.layoutManager is LinearLayoutManager) {
            lastVisiblePosition = (recyclerView.layoutManager as LinearLayoutManager).findLastVisibleItemPosition()
            firstVisiblePosition = (recyclerView.layoutManager as LinearLayoutManager).findFirstVisibleItemPosition()
//            println("$firstVisiblePosition - $lastVisiblePosition" )
        } else {
            lastVisiblePosition = -1
            firstVisiblePosition = -1
        }
    }

}

class StateHolder(textView: TextView): RecyclerView.ViewHolder(textView)

class AdapterWrapper<H: RecyclerView.ViewHolder, in T: RecyclerView.Adapter<H>>(private val delegate: T): RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val HIDE_STATE = 0
    private val SHOW_REFRESH = 1
    private val SHOW_LOAD_MORE = 2

    private val VIEW_TYPE_REFRESH = -999
    private val VIEW_TYPE_LOAD_MORE = -998
    private val VIEW_ITEM = -997

    var state = HIDE_STATE

    fun showRefreshView(show: Boolean) {
        state = if (show) SHOW_REFRESH else HIDE_STATE
        notifyItemChanged(0)
    }

    fun showLoadMoreView(show: Boolean) {
        state = if (show) SHOW_LOAD_MORE else HIDE_STATE
        notifyItemChanged(itemCount - 1)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == VIEW_TYPE_REFRESH) {
            val textView = TextView(parent.context)
            textView.height = 100
            textView.text = "refresh..."
            return StateHolder(textView)
        } else if (viewType == VIEW_TYPE_LOAD_MORE) {
            val textView = TextView(parent.context)
            textView.height = 100
            textView.text = "loading..."
            return StateHolder(textView)
        }
        return delegate.onCreateViewHolder(parent, viewType)
    }

    override fun getItemViewType(position: Int): Int {
        return when (position) {
            0 -> VIEW_TYPE_REFRESH
            delegate.itemCount -> VIEW_TYPE_LOAD_MORE;
            else -> VIEW_ITEM
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is StateHolder) {
            if (state == HIDE_STATE) {
                holder.itemView.visibility = RecyclerView.GONE
            } else if ((state == SHOW_LOAD_MORE && position == delegate.itemCount) || (state == SHOW_REFRESH && position == 0)) {
                holder.itemView.visibility = RecyclerView.VISIBLE
            }
        } else {
            delegate.onBindViewHolder(holder as H, position - 1)
        }
    }

    override fun getItemCount(): Int {
        return delegate.itemCount + 1
    }
}

class LoadMoreRecyclerView(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : RecyclerView(context, attrs, defStyleAttr) {

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context) : this(context, null)

    private val viewStateListener = { isRefresh: Boolean, isLoadMore: Boolean ->
        if (adapter is AdapterWrapper<*, *>) {
            val cast: AdapterWrapper<*, *> = adapter as AdapterWrapper<*, *>
            when {
                isRefresh -> cast.showRefreshView(true)
                isLoadMore -> cast.showLoadMoreView(true)
                else -> {
                    cast.showRefreshView(false)
                    cast.showLoadMoreView(false)
                }
            }
        }
    }

    private val scrollListener by lazy {
        val listener = InnerScrollListener()
        listener.viewStateListener = viewStateListener
        listener
    }

    init {
        addOnScrollListener(scrollListener)
    }

    fun setOnRefreshListener(listener: () -> Unit) {
        scrollListener.onRefreshListener = listener
    }

    fun setOnLoadMoreListener(listener: () -> Unit) {
        scrollListener.onLoadMoreListener = listener
    }

    fun loadCompleted() {
        scrollListener.isLoading = false
    }
}